### PR TITLE
BOLT2: Check we don't send and accept 0-msat HTLC

### PR DESCRIPTION
**Pull request by ariard**
Failing this requirement at sending means a strict receiver would
fail our channel while processing at HTLC routed from a third-party.

Fix by enforcing check on both sender and receiver side.

BOLT2:

A sending node MUST offer amount_msat greater than 0.

A receiving node receiving an amount_msat equal to 0, OR less than its own htlc_minimum_msat SHOULD fail the channel.

### CODE REVIEW

File: lightning/src/ln/channel.rs [hunk 1]
```diff
@@ -1597,6 +1597,9 @@ impl<ChanSigner: ChannelKeys> Channel<ChanSigner> {
 		if msg.amount_msat > self.channel_value_satoshis * 1000 {
 			return Err(ChannelError::Close("Remote side tried to send more than the total value of the channel"));
 		}
+		if msg.amount_msat == 0 {
```
Comment (TheBlueMatt) [inline lightning/src/ln/channel.rs]: Would be nice to test this line as well. Another way to do it that would be easier to test would be to just set our_htlc_minimum_msat to at least 1 always.

**Comment by ariard** [comment]
Rebased 91a9321

File: lightning/src/ln/channel.rs (hunk 1 above)
Comment (ariard) [reply lightning/src/ln/channel.rs]: Add a commit for this, without modifying default config

**Comment by TheBlueMatt** [comment]
Looks like this needs a further rebase on top of #509.

**Comment by ariard** [comment]
Rebased 3d6b2d9

File: lightning/src/ln/channel.rs [hunk 2]
```diff
@@ -1652,6 +1652,9 @@ impl<ChanSigner: ChannelKeys> Channel<ChanSigner> {
...
+			return Err(ChannelError::Ignore("Remote side tried to send a 0-msat HTLC"));
```
Comment (TheBlueMatt) [inline lightning/src/ln/channel.rs]: I dont think this is right? We should either ::Close() or we should add the HTLC and fail it backwards as soon as they commit.

Comment (ariard) [reply lightning/src/ln/channel.rs]: You right, should fail the channel IIRC the spec, will update

**Comment by ariard** [comment]
Updated at cf2918f with rebase errors fixed, ::Ignore() switched for a ::Close() and receiver-test updated to check for channel closing.

**Comment by TheBlueMatt** [comment]
I believe this needs further updates (and a rebase onto master to make travis pass).

**Comment by ariard** [comment]
Rebased 16edc6d

**Comment by TheBlueMatt** [comment]
Doesn't build.

**Comment by TheBlueMatt** [comment]
As discussed, we should never be sending an htlc_minimum_msat of < 1 to any of our peers, so this needs an update to check for that.

File: lightning/src/util/config.rs [hunk 3]
```diff
@@ -51,13 +51,21 @@ pub struct ChannelHandshakeConfig {
 	/// Default value: BREAKDOWN_TIMEOUT (currently 144), we enforce it as a minimum at channel
 	/// opening so you can tweak config to ask for more security, not less.
 	pub our_to_self_delay: u16,
+	/// Set to the smallest value HTLC we will accept to process.
+	///
+	/// Announced in `open_channel`/`accept_channel` according if it's an inbound/outbound
+	/// channel. Enforced at `update_add_htlc` reception with other HTLC sanitization checks.
```
Comment (TheBlueMatt) [inline lightning/src/util/config.rs]:
```suggestion
	/// This value is sent to our counterparty on channel-open and we close the channel any time
	/// our counterparty misbehaves by sending us an HTLC with a value smaller than this.
	///
	/// If the value is less than 1, it is ignored and set to 1, as is required by the protocol.
```

Comment (ariard) [reply lightning/src/util/config.rs]: Thanks, took your suggestion

**Pull request merged**